dbt/adapters/athena/impl.py

```diff
@@ -152,20 +131,74 @@ def clean_up_table(self, database_name: str, table_name: str):
-                s3_resource = client.session.resource("s3", region_name=client.region_name, config=get_boto3_config())
-                s3_bucket = s3_resource.Bucket(bucket_name)
-                s3_bucket.objects.filter(Prefix=prefix).delete()
+            s3_location = table["Table"]["StorageDescriptor"]["Location"]
+            logger.debug(f"Deleting table data from '{s3_location}'")
+            self._delete_from_s3(client, s3_location)
+
+    @available
+    def prune_s3_table_location(self, s3_table_location: str):
+        """
+        Prunes an s3 table location.
+        This is ncessary resolve the HIVE_PARTITION_ALREADY_EXISTS error
+        that occurs during retrying after receiving a 503 Slow Down error
+        during a CTA command, if partial files have already been written to s3.
+        """
+        conn = self.connections.get_thread_connection()
+        client = conn.handle
+        self._delete_from_s3(client, s3_table_location)
 
     @available
     def quote_seed_column(self, column: str, quote_config: Optional[bool]) -> str:
         return super().quote_seed_column(column, False)
 
+    def _delete_from_s3(self, client: Any, s3_path: str):
+        """
+        Deletes files from s3.
+        Additionally parses the response from the s3 delete request and raises
+        a RunTimeException in case it included errors.
+        """
+        bucket_name, prefix = self._parse_s3_path(s3_path)
+        if self._s3_path_exists(client, bucket_name, prefix):
+            s3_resource = client.session.resource("s3", region_name=client.region_name, config=get_boto3_config())
+            s3_bucket = s3_resource.Bucket(bucket_name)
+            response = s3_bucket.objects.filter(Prefix=prefix).delete()
+            is_all_successful = True
+            for res in response:
+                if "Errors" in res:
+                    for err in res["Errors"]:
+                        is_all_successful = False
+                        logger.error(
+                            "Failed to delete files: Key='{}', Code='{}', Message='{}', s3_bucket='{}'",
+                            err["Key"],
+                            err["Code"],
+                            err["Message"],
+                            bucket_name,
+                        )
+            if is_all_successful is False:
+                raise RuntimeException("Failed to delete files from S3.")
+        else:
+            logger.debug("S3 path does not exist")
+
+    @staticmethod
+    def _parse_s3_path(s3_path: str) -> Tuple[str, str]:
+        """
+        Parses and splits an s3 path into bucket name and prefix.
+        This assumes that s3_path is a prefix instead of a URI. It adds a
+        trailing slash to the prefix, if there is none.
+        """
+        o = urlparse(s3_path, allow_fragments=False)
+        bucket_name = o.netloc
+        prefix = o.path.lstrip("/").rstrip("/") + "/"
+        return bucket_name, prefix
```

Comment:
Could you add bunch of unit tests for this method?

Reply:
added!